File: backend/python_backup/routes/jobs.py
from flask import Blueprint, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity
from models import Resume

jobs_bp = Blueprint('jobs', __name__)

# Mock database of current job listings
JOB_LISTINGS = [
    {
        'id': 'job-1',
        'title': 'React Front-End Developer',
        'company': 'TechCorp Systems',
        'location': 'Bengaluru (Hybrid)',
        'type': 'Full-time',
        'salary': '₹8,00,000 - ₹12,00,000/yr',
        'keywords': ['react', 'javascript', 'tailwind', 'typescript', 'redux', 'html', 'css'],
        'description': 'Responsible for developing highly responsive web interfaces using React.js and modern CSS frameworks.'
    },
    {
        'id': 'job-2',
        'title': 'Flask / Python Backend Developer',
        'company': 'DevScale Solutions',
        'location': 'Remote',
        'type': 'Full-time',
        'salary': '₹10,00,000 - ₹15,00,000/yr',
        'keywords': ['python', 'flask', 'sql', 'postgresql', 'apis', 'jwt', 'docker', 'rest'],
        'description': 'Build high performance REST APIs and manage relational database schemas using SQLAlchemy.'
    },
    {
        'id': 'job-3',
        'title': 'Data Analyst',
        'company': 'FinanceFlow Analytics',
        'location': 'Mumbai',
        'type': 'Full-time',
        'salary': '₹7,00,000 - ₹10,00,000/yr',
        'keywords': ['sql', 'python', 'pandas', 'tableau', 'excel', 'powerbi', 'statistics'],
        'description': 'Translate raw operational financial data into actionable visual insights and business intelligence.'
    },
    {
        'id': 'job-4',
        'title': 'Machine Learning Intern',
        'company': 'DeepAI Lab',
        'location': 'Remote',
        'type': 'Internship',
        'salary': '₹25,000/mo',
        'keywords': ['python', 'pytorch', 'tensorflow', 'machine learning', 'numpy', 'scikit-learn'],
        'description': 'Train, validate and deploy neural networks and custom models for image classification and NLP.'
    },
    {
        'id': 'job-5',
        'title': 'UI/UX Product Designer',
        'company': 'CreativeHub Agency',
        'location': 'Pune',
        'type': 'Full-time',
        'salary': '₹6,00,000 - ₹9,00,000/yr',
        'keywords': ['figma', 'design', 'wireframe', 'prototype', 'adobe', 'ux', 'ui'],
        'description': 'Design user-centric interfaces and map complex user journeys for Web/Mobile products.'
    },
    {
        'id': 'job-6',
        'title': 'Business Analyst (MBA Fresher)',
        'company': 'MarketGrow Consultancy',
        'location': 'New Delhi',
        'type': 'Full-time',
        'salary': '₹9,00,000 - ₹13,00,000/yr',
        'keywords': ['management', 'strategy', 'analysis', 'communication', 'agile', 'scrum', 'sql'],
        'description': 'Coordinate requirements between product stakeholders and technical engineering leads.'
    }
]

@jobs_bp.route('/match/<resume_id>', methods=['GET'])
@jwt_required()
def match_jobs(resume_id):
    user_id = get_jwt_identity()
    resume = Resume.query.get(resume_id)
    
    if not resume:
        return jsonify({'error': 'Resume not found'}), 404
        
    if resume.user_id != int(user_id):
        return jsonify({'error': 'Unauthorized'}), 403
        
    # Extract keywords from resume
    content = resume.content or {}
    skills = [s.lower() for s in content.get('skills', [])]
    
    # Also extract from experience/projects text to improve match
    experience_desc = " ".join([exp.get('description', '').lower() for exp in content.get('experience', [])])
    projects_desc = " ".join([proj.get('description', '').lower() for proj in content.get('projects', [])])
    
    resume_corpus = " ".join(skills) + " " + experience_desc + " " + projects_desc
    
    # Score each job
    matches = []
    for job in JOB_LISTINGS:
        matching_keywords = []
        for kw in job['keywords']:
            # check if keyword matches skills or project/experience description
            if kw in resume_corpus:
                matching_keywords.append(kw)
                
        # calculate percentage match
        kw_len = len(job['keywords'])
        match_score = int((len(matching_keywords) / kw_len) * 100) if kw_len > 0 else 0
        
        # Give fallback 35% minimum match if they have at least 1 keyword, or 20% flat
        if match_score == 0 and len(matching_keywords) > 0:
            match_score = 35
            
        matches.append({
            'job': job,
            'match_score': max(20, match_score), # min 20% to look realistic
            'matching_keywords': matching_keywords
        })
        
    # Sort matches by match_score desc
    matches.sort(key=lambda x: x['match_score'], reverse=True)
    
    return jsonify(matches), 200

File: backend/routes/jobs.js
import express from 'express';
import { Resume } from '../models/Resume.js';
import { protect } from '../middleware/auth.js';
import { ForbiddenError, NotFoundError } from '../utils/errors.js';
import { mlClient } from '../utils/mlClient.js';

const router = express.Router();

// Mock database of current job listings
const JOB_LISTINGS = [
  {
    id: 'job-1',
    title: 'React Front-End Developer',
    company: 'TechCorp Systems',
    location: 'Bengaluru (Hybrid)',
    type: 'Full-time',
    salary: '₹8,00,000 - ₹12,00,000/yr',
    keywords: ['react', 'javascript', 'tailwind', 'typescript', 'redux', 'html', 'css'],
    description: 'Responsible for developing highly responsive web interfaces using React.js and modern CSS frameworks.'
  },
  {
    id: 'job-2',
    title: 'Flask / Python Backend Developer',
    company: 'DevScale Solutions',
    location: 'Remote',
    type: 'Full-time',
    salary: '₹10,00,000 - ₹15,00,000/yr',
    keywords: ['python', 'flask', 'sql', 'postgresql', 'apis', 'jwt', 'docker', 'rest'],
    description: 'Build high performance REST APIs and manage relational database schemas using SQLAlchemy.'
  },
  {
    id: 'job-3',
    title: 'Data Analyst',
    company: 'FinanceFlow Analytics',
    location: 'Mumbai',
    type: 'Full-time',
    salary: '₹7,00,000 - ₹10,00,000/yr',
    keywords: ['sql', 'python', 'pandas', 'tableau', 'excel', 'powerbi', 'statistics'],
    description: 'Translate raw operational financial data into actionable visual insights and business intelligence.'
  },
  {
    id: 'job-4',
    title: 'Machine Learning Intern',
    company: 'DeepAI Lab',
    location: 'Remote',
    type: 'Internship',
    salary: '₹25,000/mo',
    keywords: ['python', 'pytorch', 'tensorflow', 'machine learning', 'numpy', 'scikit-learn'],
    description: 'Train, validate and deploy neural networks and custom models for image classification and NLP.'
  },
  {
    id: 'job-5',
    title: 'UI/UX Product Designer',
    company: 'CreativeHub Agency',
    location: 'Pune',
    type: 'Full-time',
    salary: '₹6,00,000 - ₹9,00,000/yr',
    keywords: ['figma', 'design', 'wireframe', 'prototype', 'adobe', 'ux', 'ui'],
    description: 'Design user-centric interfaces and map complex user journeys for Web/Mobile products.'
  },
  {
    id: 'job-6',
    title: 'Business Analyst (MBA Fresher)',
    company: 'MarketGrow Consultancy',
    location: 'New Delhi',
    type: 'Full-time',
    salary: '₹9,00,000 - ₹13,00,000/yr',
    keywords: ['management', 'strategy', 'analysis', 'communication', 'agile', 'scrum', 'sql'],
    description: 'Coordinate requirements between product stakeholders and technical engineering leads.'
  }
];

// Match resume to jobs
router.get('/match/:resumeId', protect, async (req, res, next) => {
  try {
    const resume = await Resume.findById(req.params.resumeId);
    if (!resume) {
      return next(new NotFoundError('Resume not found.'));
    }

    if (resume.userId.toString() !== req.user._id.toString()) {
      return next(new ForbiddenError('Unauthorized to access matching for this resume.'));
    }

    // Extract keywords from resume content
    const content = resume.content || {};
    const personal = content.personal || {};
    
    // Extracted skills
    let skillsList = [];
    const skills = content.skills || {};
    if (typeof skills === 'object') {
      skillsList = Object.values(skills).flatMap(s => String(s).toLowerCase().split(/[\s,]+/));
    } else if (Array.isArray(skills)) {
      skillsList = skills.map(s => String(s).toLowerCase());
    }

    // Extracted work experience descriptions
    const experienceList = (content.experience || []).map(exp => (exp.description || '').toLowerCase());
    const projectsList = (content.projects || []).map(proj => (proj.description || '').toLowerCase());

    const resumeCorpus = [
      ...skillsList,
      ...experienceList,
      ...projectsList,
      (personal.summary || '').toLowerCase()
    ].join(' ');

    // Try Python ML Service Matching (Cosine Similarity) first
    const mlMatches = await mlClient.matchJobs(resumeCorpus, JOB_LISTINGS);
    
    if (mlMatches) {
      // Map ML results back to job details
      const enrichedMatches = mlMatches.map(m => {
        const job = JOB_LISTINGS.find(j => j.id === m.job_id);
        return {
          job,
          match_score: m.match_score,
          matching_keywords: m.matching_keywords,
          algorithm: 'Cosine Similarity (TF-IDF)'
        };
      });
      return res.status(200).json(enrichedMatches);
    }

    // Fallback: Local Regex Keyword Match
    const matches = JOB_LISTINGS.map(job => {
      const matchingKeywords = [];
      for (const kw of job.keywords) {
        if (resumeCorpus.includes(kw.toLowerCase())) {
          matchingKeywords.push(kw);
        }
      }

      // Calculate matching ratio percentage
      const kwLength = job.keywords.length;
      let matchScore = kwLength > 0 ? Math.floor((matchingKeywords.length / kwLength) * 100) : 0;

      // Provide realistic minimum match fallback if at least one keyword is matched
      if (matchScore === 0 && matchingKeywords.length > 0) {
        matchScore = 35;
      }

      return {
        job,
        match_score: Math.max(20, matchScore), // Minimum 20% to look realistic
        matching_keywords: matchingKeywords,
        algorithm: 'Regex Keyword Match (Fallback)'
      };
    });

    // Sort descending by match score
    matches.sort((a, b) => b.match_score - a.match_score);

    res.status(200).json(matches);
  } catch (error) {
    next(error);
  }
});

export default router;
